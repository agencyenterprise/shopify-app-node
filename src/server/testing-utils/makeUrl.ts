import http from 'http';
import listen from 'test-listen';
import Koa, { ExtendableContext } from 'koa';

const app = new Koa();

export const listeners = new Map();

export default async function makeUrl(fn: (ctx: ExtendableContext) => void) {
  const listener = listeners.get(fn);
  if (listener) {
    return listener.url;
  }

  app.use(fn);
  const server = new http.Server(app.callback());
  const url = await listen(server);
  listeners.set(fn, {
    url,
    server,
  })
  return url;
}

