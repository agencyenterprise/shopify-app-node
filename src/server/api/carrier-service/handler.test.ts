import handler from 'src/server/api/carrier-service/handler'
import makeUrl from 'src/server/testing-utils/makeUrl';
import axios from 'axios';
import { rateRequest1 } from 'src/server/api/carrier-service/__fixtures__/rateRequests';

describe('handler', () => {
  it('works', async () => {

    const url = await makeUrl(handler);
    const { data } = await axios.post(url, rateRequest1)
    expect(data).toMatchSnapshot();

  })
})
