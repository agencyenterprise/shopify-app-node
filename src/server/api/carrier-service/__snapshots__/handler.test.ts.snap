// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler works 1`] = `
Object {
  "rates": Array [
    Object {
      "currency": "CAD",
      "description": "This is the fastest option by far",
      "max_delivery_date": "2013-04-12 14:48:45 -0400",
      "min_delivery_date": "2013-04-12 14:48:45 -0400",
      "service_code": "ON",
      "service_name": "canadapost-overnight",
      "total_price": "1295",
    },
    Object {
      "currency": "USD",
      "max_delivery_date": "2013-04-12 14:48:45 -0400",
      "min_delivery_date": "2013-04-12 14:48:45 -0400",
      "service_code": "2D",
      "service_name": "fedex-2dayground",
      "total_price": "2934",
    },
    Object {
      "currency": "USD",
      "max_delivery_date": "2013-04-12 14:48:45 -0400",
      "min_delivery_date": "2013-04-12 14:48:45 -0400",
      "service_code": "1D",
      "service_name": "fedex-priorityovernight",
      "total_price": "3587",
    },
  ],
}
`;
